File: src/apple-ai.swift
import Foundation
import FoundationModels

// MARK: - C-compatible data structures

@available(macOS 26.0, *)

@_cdecl("apple_ai_init")
public func appleAIInit() -> Bool {
    // Initialize and return success status
    return true
}

@_cdecl("apple_ai_check_availability")
public func appleAICheckAvailability() -> Int32 {
    let model = SystemLanguageModel.default
    let availability = model.availability
    
    switch availability {
    case .available:
        return 1 // Available
    case .unavailable(let reason):
        switch reason {
        case .deviceNotEligible:
            return -1 // Device not eligible
        case .appleIntelligenceNotEnabled:
            return -2 // Apple Intelligence not enabled
        case .modelNotReady:
            return -3 // Model not ready
        @unknown default:
            return -99 // Unknown error
        }
    @unknown default:
        return -99 // Unknown error
    }
}

@_cdecl("apple_ai_get_availability_reason")
public func appleAIGetAvailabilityReason() -> UnsafeMutablePointer<CChar>? {
    let model = SystemLanguageModel.default
    let availability = model.availability
    
    switch availability {
    case .available:
        return strdup("Model is available")
    case .unavailable(let reason):
        let reasonString: String
        switch reason {
        case .deviceNotEligible:
            reasonString = "Device not eligible for Apple Intelligence. Supported devices: iPhone 15 Pro/Pro Max or newer, iPad with M1 chip or newer, Mac with Apple Silicon"
        case .appleIntelligenceNotEnabled:
            reasonString = "Apple Intelligence not enabled. Enable it in Settings > Apple Intelligence & Siri"
        case .modelNotReady:
            reasonString = "AI model not ready. Models are downloaded automatically based on network status, battery level, and system load. Please wait and try again later."
        @unknown default:
            reasonString = "Unknown availability issue"
        }
        return strdup(reasonString)
    @unknown default:
        return strdup("Unknown availability status")
    }
}

@_cdecl("apple_ai_get_supported_languages_count")
public func appleAIGetSupportedLanguagesCount() -> Int32 {
    let model = SystemLanguageModel.default
    return Int32(Array(model.supportedLanguages).count)
}

@_cdecl("apple_ai_get_supported_language")
public func appleAIGetSupportedLanguage(index: Int32) -> UnsafeMutablePointer<CChar>? {
    let model = SystemLanguageModel.default
    let languagesArray = Array(model.supportedLanguages)
    
    guard index >= 0 && index < Int32(languagesArray.count) else {
        return nil
    }
    
    let language = languagesArray[Int(index)]
    let locale = Locale(identifier: language.maximalIdentifier)
    
    // Get the display name in the current locale
    if let displayName = locale.localizedString(forIdentifier: language.maximalIdentifier) {
        return strdup(displayName)
    }
    
    // Fallback to language code if display name is not available
    if let languageCode = language.languageCode?.identifier {
        return strdup(languageCode)
    }
    
    return strdup("Unknown")
}

@_cdecl("apple_ai_generate_response")
public func appleAIGenerateResponse(
    prompt: UnsafePointer<CChar>,
    temperature: Double,
    maxTokens: Int32
) -> UnsafeMutablePointer<CChar>? {
    let promptString = String(cString: prompt)
    
    // Use semaphore to convert async to sync
    let semaphore = DispatchSemaphore(value: 0)
    var result: String = "Error: No response"
    
    Task {
        do {
            let model = SystemLanguageModel.default
            
            // Check availability first
            let availability = model.availability
            guard case .available = availability else {
                result = "Error: Apple Intelligence not available"
                semaphore.signal()
                return
            }
            
            // Create session
            let session = LanguageModelSession()
            
            // Create generation options
            var options = GenerationOptions()
            if temperature > 0 {
                options = GenerationOptions(temperature: temperature, maximumResponseTokens: maxTokens > 0 ? Int(maxTokens) : nil)
            } else if maxTokens > 0 {
                options = GenerationOptions(maximumResponseTokens: Int(maxTokens))
            }
            
            // Generate response
            let response = try await session.respond(to: promptString, options: options)
            result = response.content
            
        } catch {
            result = "Error: \(error.localizedDescription)"
        }
        
        semaphore.signal()
    }
    
    // Wait for async operation to complete
    semaphore.wait()
    
    return strdup(result)
}

@_cdecl("apple_ai_generate_response_with_history")
public func appleAIGenerateResponseWithHistory(
    messagesJson: UnsafePointer<CChar>,
    temperature: Double,
    maxTokens: Int32
) -> UnsafeMutablePointer<CChar>? {
    let messagesJsonString = String(cString: messagesJson)
    
    // Use semaphore to convert async to sync
    let semaphore = DispatchSemaphore(value: 0)
    var result: String = "Error: No response"
    
    Task {
        do {
            let model = SystemLanguageModel.default
            
            // Check availability first
            let availability = model.availability
            guard case .available = availability else {
                result = "Error: Apple Intelligence not available"
                semaphore.signal()
                return
            }
            
            // Parse messages from JSON
            guard let messagesData = messagesJsonString.data(using: .utf8) else {
                result = "Error: Invalid JSON data"
                semaphore.signal()
                return
            }
            
            let messages = try JSONDecoder().decode([ChatMessage].self, from: messagesData)
            
            guard !messages.isEmpty else {
                result = "Error: No messages provided"
                semaphore.signal()
                return
            }
            
            // Get the last message as the current prompt
            let lastMessage = messages.last!
            let currentPrompt = lastMessage.content
            
            // Convert previous messages to transcript for conversation context
            let previousMessages = messages.count > 1 ? Array(messages.dropLast()) : []
            let transcriptEntries = convertMessagesToTranscript(previousMessages)
            
            // Create session with conversation history
            let transcript = Transcript(entries: transcriptEntries)
            let session = LanguageModelSession(transcript: transcript)
            
            // Create generation options
            var options = GenerationOptions()
            if temperature > 0 {
                options = GenerationOptions(temperature: temperature, maximumResponseTokens: maxTokens > 0 ? Int(maxTokens) : nil)
            } else if maxTokens > 0 {
                options = GenerationOptions(maximumResponseTokens: Int(maxTokens))
            }
            
            // Generate response using the current prompt with history
            let response = try await session.respond(to: currentPrompt, options: options)
            result = response.content
            
        } catch {
            result = "Error: \(error.localizedDescription)"
        }
        
        semaphore.signal()
    }
    
    // Wait for async operation to complete
    semaphore.wait()
    
    return strdup(result)
}

@_cdecl("apple_ai_free_string")
public func appleAIFreeString(ptr: UnsafeMutablePointer<CChar>?) {
    if let ptr = ptr {
        free(ptr)
    }
}

// MARK: - Helper functions

private struct ChatMessage: Codable {
    let role: String
    let content: String
    let name: String?
    
    init(role: String, content: String, name: String? = nil) {
        self.role = role
        self.content = content
        self.name = name
    }
}

private func convertMessagesToTranscript(_ messages: [ChatMessage]) -> [Transcript.Entry] {
    var entries: [Transcript.Entry] = []
    
    // Process all messages in order
    for message in messages {
        let textSegment = Transcript.TextSegment(content: message.content)
        
        switch message.role.lowercased() {
        case "system":
            // Convert system messages to instructions
            let instructions = Transcript.Instructions(
                segments: [.text(textSegment)],
                toolDefinitions: []
            )
            entries.append(.instructions(instructions))
            
        case "user":
            // Convert user messages to prompts
            let prompt = Transcript.Prompt(
                segments: [.text(textSegment)]
            )
            entries.append(.prompt(prompt))
            
        case "assistant":
            // Convert assistant messages to responses
            let response = Transcript.Response(
                assetIDs: [],
                segments: [.text(textSegment)]
            )
            entries.append(.response(response))
            
        default:
            // Treat unknown roles as user messages
            let prompt = Transcript.Prompt(
                segments: [.text(textSegment)]
            )
            entries.append(.prompt(prompt))
        }
    }
    
    return entries
}

@available(macOS 26.0, *)
@_cdecl("apple_ai_generate_response_stream")
public func appleAIGenerateResponseStream(
    _ prompt: UnsafePointer<CChar>,
    _ temperature: Double,
    _ maxTokens: Int32,
    _ onChunk: (@convention(c) (UnsafePointer<CChar>?) -> Void)
) {
    let promptString = String(cString: prompt)

    Task.detached {
        do {
            let model = SystemLanguageModel.default
            guard case .available = model.availability else {
                emitError("Model unavailable", to: onChunk)
                return
            }

            var options = GenerationOptions()
            if temperature != 0.0 { options.temperature = temperature }
            if maxTokens > 0 { options.maximumResponseTokens = Int(maxTokens) }

            let session = LanguageModelSession(model: model)

            var prev = ""
            for try await cumulative in session.streamResponse(to: promptString, options: options) {
                let delta = String(cumulative.dropFirst(prev.count))
                prev = cumulative
                guard !delta.isEmpty, delta.first != ERROR_SENTINEL else { continue }

                delta.withCString { cStr in
                    onChunk(strdup(cStr))
                }
            }
            onChunk(nil)        // stream finished
        } catch {
            emitError(error.localizedDescription, to: onChunk)
        }
    }
}

// Control-B (0x02) sentinel prefix marks an error string in streaming callbacks
private let ERROR_SENTINEL: Character = "\u{0002}"

@inline(__always)
private func emitError(_ message: String, to onChunk: (@convention(c) (UnsafePointer<CChar>?) -> Void)) {
    let full = String(ERROR_SENTINEL) + message
    full.withCString { cStr in
        onChunk(strdup(cStr))
    }
} 

File: src/apple-ai.ts
import { getNativeModule } from "./native-loader";

// Initialize native module using robust loader
const native = getNativeModule();

// Types for our Apple AI library
export interface ChatMessage {
  role: "system" | "user" | "assistant";
  content: string;
  name?: string;
}

export interface GenerationOptions {
  temperature?: number;
  maxTokens?: number;
}

export interface ModelAvailability {
  available: boolean;
  reason: string;
}

// OpenAI-compatible response types
export interface ChatCompletionChunk {
  id: string;
  object: "chat.completion.chunk";
  created: number;
  model: string;
  choices: {
    index: number;
    delta: {
      role?: "assistant";
      content?: string;
    };
    finish_reason: string | null;
  }[];
}

export interface ChatCompletionResponse {
  id: string;
  object: "chat.completion";
  created: number;
  model: string;
  choices: {
    index: number;
    message: ChatMessage;
    finish_reason: string;
  }[];
}

/**
 * Apple AI library for accessing on-device foundation models
 */
export class AppleAISDK {
  /** Check availability of Apple Intelligence */
  async checkAvailability(): Promise<ModelAvailability> {
    return native.checkAvailability();
  }

  /** Get supported languages */
  getSupportedLanguages(): string[] {
    return native.getSupportedLanguages();
  }

  /** Generate a response for a prompt */
  async generateResponse(
    prompt: string,
    options: GenerationOptions = {}
  ): Promise<string> {
    return native.generateResponse(
      prompt,
      options.temperature ?? undefined,
      options.maxTokens ?? undefined
    );
  }

  /** Generate a response using conversation history */
  async generateResponseWithHistory(
    messages: ChatMessage[],
    options: GenerationOptions = {}
  ): Promise<string> {
    const messagesJson = JSON.stringify(messages);
    return native.generateResponseWithHistory(
      messagesJson,
      options.temperature ?? undefined,
      options.maxTokens ?? undefined
    );
  }

  /**
   * Stream chat completion as async generator yielding OpenAI-compatible chunks
   */
  streamChatCompletion(
    messages: ChatMessage[],
    options: GenerationOptions = {}
  ): AsyncIterableIterator<ChatCompletionChunk> {
    const completionId = `chatcmpl-${crypto.randomUUID()}`;
    const created = Math.floor(Date.now() / 1000);

    const queue: ChatCompletionChunk[] = [];
    let done = false;
    let isFirstChunk = true;

    // Pending promise controls for consumer awaiting next chunk
    let pendingResolve:
      | ((value: IteratorResult<ChatCompletionChunk>) => void)
      | null = null;
    let pendingReject: ((reason?: any) => void) | null = null;

    let error: any = null;

    // Push-based native callback
    const handleChunk = (err: any, chunk?: string | null) => {
      if (err) {
        error = err;
        done = true;
        if (pendingReject) {
          pendingReject(err);
          pendingResolve = null;
          pendingReject = null;
        }
        return;
      }

      let chatChunk: ChatCompletionChunk;

      if (chunk == null || chunk === "") {
        // Final chunk
        chatChunk = {
          id: completionId,
          object: "chat.completion.chunk",
          created,
          model: "apple-on-device",
          choices: [
            {
              index: 0,
              delta: {},
              finish_reason: "stop",
            },
          ],
        };
        done = true;
      } else {
        // Content chunk
        chatChunk = {
          id: completionId,
          object: "chat.completion.chunk",
          created,
          model: "apple-on-device",
          choices: [
            {
              index: 0,
              delta: {
                ...(isFirstChunk ? { role: "assistant" as const } : {}),
                content: chunk,
              },
              finish_reason: null,
            },
          ],
        };
        isFirstChunk = false;
      }

      // If the consumer is waiting, resolve immediately; otherwise buffer
      if (pendingResolve) {
        pendingResolve({ value: chatChunk, done: false });
        pendingResolve = null;
        pendingReject = null;
      } else {
        queue.push(chatChunk);
      }
    };

    // Use the existing streaming mechanism but with messages
    const messagesJson = JSON.stringify(messages);
    native.generateResponseStreamWithHistory?.(
      messagesJson,
      options.temperature ?? undefined,
      options.maxTokens ?? undefined,
      handleChunk
    ) ??
      (() => {
        // Fallback: if streaming with history isn't available, convert the prompt and use regular streaming
        const prompt =
          messages.map((m) => `${m.role}: ${m.content}`).join("\n") +
          "\nassistant:";
        native.generateResponseStream(
          prompt,
          options.temperature ?? undefined,
          options.maxTokens ?? undefined,
          handleChunk
        );
      })();

    return {
      next(): Promise<IteratorResult<ChatCompletionChunk>> {
        if (queue.length > 0) {
          const value = queue.shift()!;
          return Promise.resolve({ value, done: false });
        }
        if (done) {
          return Promise.resolve({ value: undefined, done: true });
        }
        if (error) {
          return Promise.reject(error);
        }
        // Wait for the next chunk
        return new Promise<IteratorResult<ChatCompletionChunk>>(
          (resolve, reject) => {
            pendingResolve = resolve;
            pendingReject = reject;
          }
        );
      },
      async return(): Promise<IteratorResult<ChatCompletionChunk>> {
        done = true;
        return { value: undefined, done: true };
      },
      async throw(err?: any): Promise<IteratorResult<ChatCompletionChunk>> {
        done = true;
        throw err;
      },
      [Symbol.asyncIterator]() {
        return this;
      },
    };
  }

  /** OpenAI-style helper with streaming support */
  createChatCompletion<T extends boolean = false>(params: {
    messages: ChatMessage[];
    temperature?: number;
    max_tokens?: number;
    model?: string;
    stream?: T;
  }): T extends true
    ? AsyncIterableIterator<ChatCompletionChunk>
    : Promise<ChatCompletionResponse>;

  // Overload for explicit non-streaming
  createChatCompletion(params: {
    messages: ChatMessage[];
    temperature?: number;
    max_tokens?: number;
    model?: string;
    stream?: false;
  }): Promise<ChatCompletionResponse>;

  // Overload for explicit streaming
  createChatCompletion(params: {
    messages: ChatMessage[];
    temperature?: number;
    max_tokens?: number;
    model?: string;
    stream: true;
  }): AsyncIterableIterator<ChatCompletionChunk>;

  // Implementation
  createChatCompletion(params: {
    messages: ChatMessage[];
    temperature?: number;
    max_tokens?: number;
    model?: string;
    stream?: boolean;
  }):
    | Promise<ChatCompletionResponse>
    | AsyncIterableIterator<ChatCompletionChunk> {
    if (params.stream === true) {
      return this.streamChatCompletion(params.messages, {
        temperature: params.temperature,
        maxTokens: params.max_tokens,
      });
    }

    // Non-streaming response
    return this.generateResponseWithHistory(params.messages, {
      temperature: params.temperature,
      maxTokens: params.max_tokens,
    }).then((content) => ({
      id: `chatcmpl-${crypto.randomUUID()}`,
      object: "chat.completion" as const,
      created: Math.floor(Date.now() / 1000),
      model: "apple-on-device",
      choices: [
        {
          index: 0,
          message: { role: "assistant" as const, content },
          finish_reason: "stop",
        },
      ],
    }));
  }

  /**
   * Stream response as async generator yielding string chunks (deltas)
   */
  streamResponse(
    prompt: string,
    options: GenerationOptions = {}
  ): AsyncIterableIterator<string> {
    const queue: string[] = [];
    let done = false;

    // Pending promise controls for consumer awaiting next chunk
    let pendingResolve: ((value: IteratorResult<string>) => void) | null = null;
    let pendingReject: ((reason?: any) => void) | null = null;

    let error: any = null;

    // Push-based native callback
    const handleChunk = (err: any, chunk?: string | null) => {
      if (err) {
        error = err;
        done = true;
        if (pendingReject) {
          pendingReject(err);
          pendingResolve = null;
          pendingReject = null;
        }
        return;
      }

      if (chunk == null || chunk === "") {
        done = true;
        if (pendingResolve) {
          pendingResolve({ value: undefined, done: true });
          pendingResolve = null;
        }
        return;
      }

      // If the consumer is waiting, resolve immediately; otherwise buffer
      if (pendingResolve) {
        pendingResolve({ value: chunk, done: false });
        pendingResolve = null;
        pendingReject = null;
      } else {
        queue.push(chunk);
      }
    };

    native.generateResponseStream(
      prompt,
      options.temperature ?? undefined,
      options.maxTokens ?? undefined,
      handleChunk
    );

    return {
      next(): Promise<IteratorResult<string>> {
        if (queue.length > 0) {
          const value = queue.shift()!;
          return Promise.resolve({ value, done: false });
        }
        if (done) {
          return Promise.resolve({ value: undefined, done: true });
        }
        if (error) {
          return Promise.reject(error);
        }
        // Wait for the next chunk
        return new Promise<IteratorResult<string>>((resolve, reject) => {
          pendingResolve = resolve;
          pendingReject = reject;
        });
      },
      async return(): Promise<IteratorResult<string>> {
        done = true;
        return { value: undefined, done: true };
      },
      async throw(err?: any): Promise<IteratorResult<string>> {
        done = true;
        throw err;
      },
      [Symbol.asyncIterator]() {
        return this;
      },
    };
  }
}

export const appleAISDK = new AppleAISDK();
